cognite/client/data_classes/datapoints.py

```diff
@@ -313,91 +581,150 @@ def _repr_html_(self) -> str:
+
+        for id, dps_arr in ids.items():
+            if id in dupe_ids:
+                id_dct[id].append(dps_arr)
+
+        for xid, dps_arr in xids.items():
+            if xid in dupe_xids:
+                xid_dct[xid].append(dps_arr)
+
+        self._id_to_item.update(id_dct)
+        self._external_id_to_item.update(xid_dct)
+
+    def get(  # type: ignore [override]
+        self,
+        id: int = None,
+        external_id: str = None,
+    ) -> Union[None, DatapointsArray, List[DatapointsArray]]:
+        # TODO: Question, can we type annotate without specifying the function?
+        return super().get(id, external_id)  # type: ignore [return-value]
+
+    def __str__(self) -> str:
+        return json.dumps(self.dump(convert_timestamps=True), indent=4)
+
+    def _repr_html_(self) -> str:
+        return self.to_pandas()._repr_html_()
+
+    def to_pandas(  # type: ignore [override]
+        self, column_names: Literal["id", "external_id"] = "external_id", include_aggregate_name: bool = True
+    ) -> "pandas.DataFrame":
+        pd = cast(Any, local_import("pandas"))
+        if dfs := [arr.to_pandas(column_names, include_aggregate_name) for arr in self.data]:
+            return pd.concat(dfs, axis="columns")
+        return pd.DataFrame(index=pd.to_datetime([]))
+
+    def dump(self, camel_case: bool = False, convert_timestamps: bool = False) -> List[Dict[str, Any]]:
+        """Dump the instance into a json serializable Python data type.
+
+        Args:
+            camel_case (bool): Use camelCase for attribute names. Default: False.
+            convert_timestamps (bool): Convert integer timestamps to ISO 8601 formatted strings. Default: False.
+
+        Returns:
+            List[Dict[str, Any]]: A list of dicts representing the instance.
+        """
+        return [dps.dump(camel_case, convert_timestamps) for dps in self.data]
+
+
 class DatapointsList(CogniteResourceList):
     _RESOURCE = Datapoints
 
+    def __init__(self, resources: Collection[Any], cognite_client: "CogniteClient" = None):
+        super().__init__(resources, cognite_client)
+
+        # Fix what happens for duplicated identifiers:
+        ids = {x.id: x for x in self.data if x.id is not None}
+        xids = {x.external_id: x for x in self.data if x.external_id is not None}
+        dupe_ids, id_dct = find_duplicates(ids), defaultdict(list)
+        dupe_xids, xid_dct = find_duplicates(xids), defaultdict(list)
+
+        for id, dps in ids.items():
+            if id in dupe_ids:
+                id_dct[id].append(dps)
+
+        for xid, dps in xids.items():
+            if xid in dupe_xids:
+                xid_dct[xid].append(dps)
+
+        self._id_to_item.update(id_dct)
+        self._external_id_to_item.update(xid_dct)
+
+    def get(  # type: ignore [override]
+        self,
+        id: int = None,
+        external_id: str = None,
+    ) -> Union[None, DatapointsList, List[DatapointsList]]:
```

Comment:
I updated this myself.

Reply:
The change is actually correct, as multiple identifiers are now allowed in the same query (to mimic the API as much as possible). This might be handy when fetching datapoints for example 50 events (when the data in between is not needed). I've already added some examples in the retrieve docstring.

The only way for this to "make sense" with the `.get` method, was to return a list of datapoints for all duplicated identifiers. I'm happy to discuss a different solution!

Reverted in `aed3a2bc`

Reply:
Then I 

> > Should we have a migration file to list all changes needed to migrate from V4 to V5?
> 
> I don't feel qualified to answer this one (whether just a change log is sufficient or not), but I'm happy to write a "migration guide" for this particular release at least!

That would be great.

Reply:
Ohhh, I see now. Thanks for claryifing.

Reply:
On second thought, I think this should always be Optional[List[DatapointList]] then.

Reply:
Just saw that it’s still wrong 🤦  Should be:
```py
Union[None, Datapoints, List[DatapointsList]]
# and not
Union[None, DatapointsList, List[DatapointsList]]
```

Reply:
`None -> []` as well?